src/sundials/sundials_context.c

```diff
@@ -18,82 +18,157 @@
+    err = SUNProfiler_Create(comm, "SUNContext Default", &profiler); 
+    SUNCheckCallNoRet(err); if (err) { break; }
 #endif
 
-  *sunctx = NULL;
-  *sunctx = (SUNContext)malloc(sizeof(struct _SUNContext));
+    err = SUNErrHandler_Create(SUNLogErrHandlerFn, NULL, &eh);
+    SUNCheckCallNoRet(err); if (err) { break; }
 
-  if (*sunctx == NULL)
-  {
+    sunctx->logger        = logger;
+    sunctx->own_logger    = logger != NULL;
+    sunctx->profiler      = profiler;
+    sunctx->own_profiler  = profiler != NULL;
+    sunctx->last_err      = 0;
+    sunctx->err_handler   = eh;
+    sunctx->comm          = comm;
+  } while (0);
+
+  if (err) {
 #if defined(SUNDIALS_BUILD_WITH_PROFILING) && !defined(SUNDIALS_CALIPER_ENABLED)
-    SUNProfiler_Free(&profiler);
+    SUNCheckCallNoRet(SUNProfiler_Free(&profiler));
 #endif
-    SUNLogger_Destroy(&logger);
-    return (-1);
+    SUNCheckCallNoRet(SUNLogger_Destroy(&logger));
+    free(sunctx);
+  } else {
+    *sunctx_out = sunctx;
   }
 
-  (*sunctx)->logger       = logger;
-  (*sunctx)->own_logger   = logger != NULL;
-  (*sunctx)->profiler     = profiler;
-  (*sunctx)->own_profiler = profiler != NULL;
+  return err;
+}
+
+SUNErrCode SUNContext_GetLastError(SUNContext sunctx)
+{
+  if (!sunctx) { return SUN_ERR_SUNCTX_CORRUPT; }
 
-  return (0);
+  SUNFunctionBegin(sunctx);
+  return sunctx->last_err;
 }
 
-int SUNContext_GetProfiler(SUNContext sunctx, SUNProfiler* profiler)
+SUNErrCode SUNContext_PeekLastError(SUNContext sunctx)
 {
-  if (sunctx == NULL)
+  if (!sunctx) { return SUN_ERR_SUNCTX_CORRUPT; }
+
+  SUNFunctionBegin(sunctx);
+  return sunctx->last_err;
+}
+
+SUNErrCode SUNContext_PushErrHandler(SUNContext sunctx, SUNErrHandlerFn err_fn, void* err_user_data)
+{
+  if (!sunctx) { return SUN_ERR_SUNCTX_CORRUPT; }
+
+  SUNFunctionBegin(sunctx);
+  SUNErrHandler new_err_handler = NULL;
+  if (SUNErrHandler_Create(err_fn, err_user_data, &new_err_handler)) {
+    return SUN_ERR_CORRUPT;
+  }
+  new_err_handler->previous = sunctx->err_handler;
+  sunctx->err_handler = new_err_handler;
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNContext_PopErrHandler(SUNContext sunctx)
+{
+  if (!sunctx) { return SUN_ERR_SUNCTX_CORRUPT; }
+
+  SUNFunctionBegin(sunctx);
+  if (sunctx->err_handler) {
+    SUNErrHandler eh = sunctx->err_handler;
+    if (sunctx->err_handler->previous) {
+      sunctx->err_handler = sunctx->err_handler->previous;
+    } else {
+      sunctx->err_handler = NULL;
+    }
+    SUNErrHandler_Destroy(eh);
+  }
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNContext_ClearErrHandlers(SUNContext sunctx)
+{
+  if (!sunctx) { return SUN_ERR_SUNCTX_CORRUPT; }
+
+  SUNFunctionBegin(sunctx);
+  while (sunctx->err_handler != NULL)
   {
-    return (-1);
+    SUNCheckCall(SUNContext_PopErrHandler(sunctx));
   }
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNContext_GetProfiler(SUNContext sunctx, SUNProfiler* profiler)
+{
+  if (!sunctx) { return SUN_ERR_SUNCTX_CORRUPT; }
+
+  SUNFunctionBegin(sunctx);
```

Comment:
No, it is not (currently). Should we have it anyways though, to follow the "rule"?

Reply:
I think it makes sense to follow the rule since it won't do any harm and if we ever use `SUNFunctionBegin` for additional purposes it will already be there.